building-an-api (#7)
Ada-Developers-Academy/core-unit-2: building-an-api/create-and-read.md

| `title=request_body["title"], ...`                  | We use keyword arguments and access the `request_body` values to create the `Book` instance                                                                                                                            |
| `db.session.add(new_book)`                          | `db.session` is the db's way of collecting changes that need to be made. Here, we are saying we want the db to add `new_book`.                                                                                         |
| `db.session.commit()`                               | Here, we are saying we want the db to save and commit the collected changes.                                                                                                                                           |
| `return`                                            | For each endpoint, we must _return_ the HTTP response                                                                                                                                                                  |

> | `Response(...)`                                     | We instantiate a `Response` object so we can `return` it                                                                                                                                                               |

Reviewer: ```suggestion
| `make_response(...)`                                     | This function instantiates a `Response` object. A `Response` object is generally what we want to return from Flask endpoint functions.                                                                                                                                                              |
```

> | `f"Book {new_book.title} successfully created"`     | The first parameter to `Response` is the HTTP response body. Until we have more specifics requirements, let's send back a string for now.                                                                              |

Reviewer: ```suggestion
| `f"Book {new_book.title} successfully created"`     | The first parameter to `make_response()` is the HTTP response body. Until we have more specific requirements, we will send back a string.                                                                              |
```

> | `status=201`                                        | We can define the status code of the `Response` by setting the kwarg `status` to a digit.                                                                                                                              |

Reviewer: ```suggestion
| `201`                                        | We can define the status code of the `Response` by passing an integer as the 2nd argument to `make_response()`. When a 2nd argument isn't specified `200` is always the default value.                                                                                                                              |
```

### Predict HTTP Requests, Responses, and Logic

Let's consider the endpoint to create a book:

| HTTP Method | Endpoint |
| ----------- | -------- |
| `GET`       | `/books` |

No request body.

The response we want to send back is a list of JSON objects (dictionaries) with `id`, `title`, and `description`:

| Response Status | Response Body                                                                                              |
| --------------- | ---------------------------------------------------------------------------------------------------------- |
| `200 OK`        | `[{"id": 1, "title": "Fictional Book Title", "description": "A fantasy novel set in an imaginary world"}]` |

During this method, we will need to:

1. Retrieve all of the books from the db
1. Format the books data into the appropriate structure (list of dictionaries, where each dictionary has `id`, `title`, and `description`)
1. Send back a response

### Working with `jsonify` from Flask

We need to import something named `jsonify` from `flask`.

```python
from flask import jsonify
```

> `jsonify` is a Flask utility function that turns its argument into JSON. We'll use `jsonify` as a way to turn a list of book dictionaries into valid JSON.

Reviewer: I manually verified the type returned from `jsonify`.
```suggestion
`jsonify` is a Flask utility function that turns its argument into JSON. We'll use `jsonify` as a way to turn a list of book dictionaries into a `Response` object.
```

| `if request.method == "GET":`                       | Separate this functionality from the Create feature by checking the `request`'s HTTP method                            |
| `... = Book.query.all()`                            | This SQLAlchemy syntax tells `Book` to `query` for `all()` books. This method returns a _list_ of instances of `Book`. |
| `books = ...`                                       | We set the list of `Book` instances into the variable `books`                                                          |
| `for book in books:`                                | We to iterate over all books in `books` so we can collect their data and format it into a response                     |
| `books_response.append( ... )`                      | We will use the `books_response` list to hold book dictionaries                                                        |
| `{ "id": book.id, ... }`                            | This is the format of dictionary we want to send back. We'll insert the values based on the `book` we're iterating on  |

> | `jsonify(books_response)`                           | `books_response` contains a list of book dictionaries. To turn it into JSON, we pass it into `jsonify()`               |

Reviewer: ```suggestion
| `jsonify(books_response)`                           | `books_response` contains a list of book dictionaries. To turn it into a `Response` object, we pass it into `jsonify()`. This will be our common practice when returning a list of something because the `make_response` function does not handle lists.               |
```

* Used `jsonify` with `jsonify(books_response)`
* Returned this JSON list with the status code `200 OK`
* Tested this request in Postman

##### !end-options
##### !answer

* Predicted the HTTP response, request, and logic for this endpoint
* Briefly considered `jsonify`
* Briefly considered `query`
* Updated this endpoint to match `GET` requests, too
* Refactored this endpoint to check `request.method`
* Got a list of `Book` instances using `Book.query.all()`
* Iterated through `books`
* Created a list of Book dictionaries in `books_response`
* Used `jsonify` with `jsonify(books_response)`
* Returned this JSON list with the status code `200 OK`
* Tested this request in Postman

##### !end-answer
### !end-challenge
<!-- prettier-ignore-end -->

## Getting a Single Book Endpoint: Preparation

Let's consider how to implement this feature:

> As a client, I want to send a request to get one existing book, so that I can see the `id`, `title`, and `description` of the book.

### Predict HTTP Requests, Responses, and Logic

Let's consider the endpoint to create a book:

| HTTP Method | Endpoint   |
| ----------- | ---------- |
| `GET`       | `/books/1` |

> No request body.

Reviewer 2: reminder about why we don't need the request body
```suggestion
`GET` requests do not include a request body, so no additional planning around the request body is needed.
```

We want to send back a single JSON object (dictionary) with `id`, `title`, and `description`:

| Response Status | Response Body                                                                                            |
| --------------- | -------------------------------------------------------------------------------------------------------- |
| `200 OK`        | `{"id": 1, "title": "Fictional Book Title", "description": "A fantasy novel set in an imaginary world"}` |

> During this method, we will need to:

Reviewer 2: the same transitional phrase as before
```suggestion
Now that we have an idea of what our endpoint should look like, we can turn our attention to how to implement it.

Our endpoint will need to:
```

1. Read the `book_id` in the request path

> 1. Retrieve the book with the matching `book_id` from the db

Reviewer 2: ```suggestion
1. Retrieve the book with the matching `book_id` from the database
```

1. Format the book data into the appropriate structure (a single dictionary with `id`, `title`, and `description`)
1. Send back a response

## Getting a Single Book Endpoint: Code

> Let's create an entirely new route.

Reviewer 2: provide motivation for why a new route is required
```suggestion
Our new route needs to read data from the incoming request path. Our existing route doesn't do this, so we'll need to add an entirely new route to our existing `Blueprint`.
```

> This route has a **route parameter**. The route `/books/1` should give us the details for the book with `id` 1. `/books/2` should give us details for book with `id` 2, `/books/3000` should give us details for book 3000, and so on.

Reviewer 2: we haven't (quite) seen the code yet, so saying it has the parameter sounds a little strange
```suggestion
Our new route needs a **route parameter**. The route `/books/1` should give us the details for the book with `id` 1. `/books/2` should give us details for book with `id` 2, `/books/3000` should give us details for book 3000, and so on.
```

> Our new route will need to account for this.

Reviewer 2: and now I'm desperate to start a passage without the words "Our new route" 🤣 
```suggestion
Let's take a look at how our new route will account for this!
```

```python
# No new import statements...

# No modifications to the other route...

@books_bp.route("/<book_id>", methods=["GET"])
def book(book_id):
    book = Book.query.get(book_id)

    return {
        "id": book.id,
        "title": book.title,
        "description": book.description

>     }

Reviewer: I would prefer that we either stick to one way of returning responses in this first lesson or we have an aside/callout sort of situation each time we introduce a new way of returning a response. (returning a dict vs. Response() vs. something else)

```

| <div style="min-width:250px;"> Piece of Code </div> | Notes                                                                                                                                                                                                                                                                                    |
| --------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `@books_bp.route("...", methods=["GET"])`           | We are setting up a new route, so we must use the `Blueprint` decorator to define it                                                                                                                                                                                                     |

> | `"/<book_id>"`                                      | This is the `Blueprint` syntax to indicate _route parameters_. `<book_id>` shows that we are looking for a variable value (could be `1`, `2`, or `3000`). We'll use this value in the function as the variable `book_id`. The name of the route parameter can be made with any decision. |

Reviewer 2: dealing with the "any decision" phrasing
```suggestion
| `"/<book_id>"`                                      | This is the `Blueprint` syntax to indicate _route parameters_. The `<book_id>` placeholder shows that we are looking for a variable value (could be `1`, `2`, or `3000`). We'll use this value in the function as the variable `book_id`, so we should use a good, descriptive name. |
```

> | `def book(book_id):`                                | This `book` function is called whenever the HTTP request matches the decorator. The name of this function can be made with any decision. We must now configure a parameter to this method, `book_id`. This parameter name must match the route parameter in the decorator.               |

Reviewer 2: dealing with "any decision"
details about exactly what ends up in the parameter
further edits may be required if there is concern about the name collision between the function and the variable.

```suggestion
| `def book(book_id):`                                | This `book` function is called whenever the HTTP request matches the decorator. The name of this function should follow the previously discussed guidelines. We must add a parameter to this method, `book_id`. This parameter name must match the route parameter in the decorator. It will receive the part of the request path that lines up with the placeholder in the route.               |
```

| `Book.query.get(...)`                               | This is the SQLAlchemy syntax to query for one `Book` resource. This method returns an instance of `Book`.                                                                                                                                                                               |

> | `Book.query.get(book_id)`                           | We must pass in the primary key of a book here. The primary key of the book we're looking for is in the route parameter, `book_id`.                                                                                                                                                      |

Reviewer 2: ```suggestion
| `Book.query.get(book_id)`                           | We must pass in the primary key of a book here. The primary key of the book we're looking for was provided in the route parameter, `book_id`.                                                                                                                                                      |
```

> | `{ "id": book_id, ... }`                            | We can create a dictionary literal for our HTTP response.                                                                                                                                                                                                                                |

Reviewer 2: use the result book object to match the code
```suggestion
| `{ "id": book.id, ... }`                            | We can create a dictionary literal for our HTTP response.                                                                                                                                                                                                                                |
```

| `return`                                            | As always, we must return a response. Flask will default to returning status `200 OK`.                                                                                                                                                                                                   |

### !callout-info

> 

Reviewer 2: add a callout to warn against putting too much stock in the placeholder names (python doesn't know what a book id is and blindly takes whats left of the path)

```suggestion

### !callout-warning

## Python Doesn't Know What a `book_id` Is

We called the route parameter `book_id` and expect it to be an id of a book. But just like regular variable names, Python has no idea what a `book_id` is. All it does is look for whatever part of the URL path follows `/books/` and captures that into the variable `book_id`. So a request for `/books/my_long_value` would set the value of `book_id` to be `"my_long_value"`.

<br />

We should be careful to avoid thinking that Python uses the name of the parameter to do any kind of validation. As usual, the name provides information to us developers, not to Python. 

### !end-callout

```

> ## Why Didn't We Use `jsonify()` on the Dictionary?
> 
> For reasons that are less important to memorize, Flask will automatically convert a dictionary into an HTTP response body.

Reviewer: ```suggestion
## Why Didn't We Call `jsonify()` or `make_response()` on the Dictionary?

For reasons that are less important to memorize, Flask will automatically convert a dictionary into an HTTP response body. If we don't want to remember this exception, we can call `jsonify()` or `make_response()` with the dictionary as an argument to return the result.
```
